### PR TITLE
Fix #584 - ConcurrentModificationException

### LINKED ISSUE #584: ConcurrentModificationException, while spawning zombie
> 2017/10/31 17:01:57 [SCHWERWIEGEND] Error occurred while pulsing world world
java.util.ConcurrentModificationException
	at java.util.ArrayList.forEach(Unknown Source)
	at net.glowstone.entity.ai.TaskManager.pulse(TaskManager.java:66)
	at net.glowstone.entity.GlowEntity.pulse(GlowEntity.java:470)
	at net.glowstone.entity.GlowLivingEntity.pulse(GlowLivingEntity.java:184)
	at net.glowstone.entity.monster.GlowMonster.pulse(GlowMonster.java:54)
	at net.glowstone.GlowWorld.pulse(GlowWorld.java:395)
	at net.glowstone.scheduler.WorldScheduler$WorldThread.run(WorldScheduler.java:139)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)

Glowstone build(compiled locally): 2017.10.0-SNAPSHOT.6ecd65a (MC: 1.12.2)
Glowstone build(official): 2017.10.0-SNAPSHOT.18a3b49

Steps to reproduce:
1.  Spawn a zombie with a spawnegg
2. Hit the zombie(so i tried to force him to move)
3. Wait for movements
(4. Sometimes i timed out, after placing 2 zombies)

EDIT: Would someone please edit the title to "ConcurrentModificationException caused by zombie movements" or something similiar?

### PR DESCRIPTION
**Pull request by Microsamp**
We should use a CopyOnWriteArrayList to prevent ConcurrentModificationExceptions, while iterating and removing/adding some elements to the list.

Ps: I tried to implement/test ReadWriteLocks, but the performance and stability(deadlocks etc.) wasn't that great.

 https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html

### CODE REVIEW

**Comment by aramperes** [comment]
Hello @Microsamp,

Thank you for your contribution. Please make sure you sign the Contributor License Agreement before we review your pull request. See the comment left by the CLA assistant bot above.

**Comment by Microsamp** [comment]
Hello @momothereal,

i signed the Contributor License Agreement.

**Comment by aramperes** [comment]
I'm worried about the possible performance hit this would cause. Keep in mind that each entity has a Task Manager being pulsed every tick, and this solution appears to be quite costly according to the docs.

**Comment by aramperes** [comment]
What about using `ConcurrentSkipListSet`? From what I understand, it is safely iterable, but it may happen that a task won't be pulsed until the next tick after it is added.

**Comment by Microsamp** [comment]
> What about using ConcurrentSkipListSet?

Mhmmm, it seems to be a better choice for performance.

> but it may happen that a task won't be pulsed until the next tick after it is added.

I think it doesn't matter if a task is executed one tick later.

**Comment by Microsamp** [comment]
WIP

EDIT: The ConcurrentSkipListSet throwed an undocumented(https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListSet.html) exception:
>Caused by: java.lang.ClassCastException: net.glowstone.entity.ai.LookAtPlayerTask cannot be cast to java.lang.Comparable
at java.util.concurrent.ConcurrentSkipListMap.cpr(Unknown Source)
at java.util.concurrent.ConcurrentSkipListMap.doPut(Unknown Source)
at java.util.concurrent.ConcurrentSkipListMap.putIfAbsent(Unknown Source)
at java.util.concurrent.ConcurrentSkipListSet.add(Unknown Source)
at net.glowstone.entity.ai.TaskManager.addTask(TaskManager.java:85)`

I think if we want to use this Set, we have to implement Comparable to every EntityTask.

Ps: Sry for wrong push(to my fork). I accidentally used an old jar...

**Comment by aramperes** [comment]
I guess we could use the Comparable class to prioritize tasks (ordering), but that it is probably out of the scope of this PR.

**Comment by Microsamp** [comment]
The question is: Should i fix the exception with the first commit, should i prioritize all EntityTasks with 0 or should i wait for a full concept of your idea

**Comment by aramperes** [comment]
If it is outside of the scope, you can put a placeholder to fix the exception (for instance, using 0 for the Comparable impl), as long as it kinda works.

**Comment by Microsamp** [comment]
So: I implemented your workaround with ConcurrentSkipListSet and default priority(comparable#compareTo) of zero.

I spawned many zombies(20-30) and hitted them. The zombies moved around and died(diamond-sword) after 2 or 3 punches.
After killing all zombies the thread number(/glowstone) is now 18(normal).

File: src/main/java/net/glowstone/entity/ai/EntityTask.java [hunk 1]
```diff
@@ -14,6 +14,11 @@ public EntityTask(String name) {
         this.name = name;
     }
 
+    @Override
+    public int compareTo(EntityTask other) {
+        return 0;
```
Comment (aramperes) [inline src/main/java/net/glowstone/entity/ai/EntityTask.java]: Add a todo here (`// todo: AI task priority`)

**Review by aramperes — CHANGES_REQUESTED** [review]

**Review by aramperes — APPROVED** [review]

**Comment by aramperes** [comment]
Thank you for your contribution!

**Pull request merged**